tracer/src/Datadog.Tracer.Native/fault_tolerant_rewriter.cpp

```diff
@@ -0,0 +1,594 @@
+        const auto instrumentingProduct = m_methodRewriter->GetInstrumentingProduct(moduleHandler, methodHandler);
+
+        if (FaultTolerantTracker::Instance()->IsInstrumentationVersionSucceeded(moduleId, methodId, instrumentationVersion, instrumentingProduct))
+        {
+            // Request Revert for the original method
+            //std::vector<MethodIdentifier> requests = {{moduleId, methodId}};
+            //auto promise = std::make_shared<std::promise<void>>();
+            //auto future = promise->get_future();
+            //m_rejit_handler->EnqueueRequestRevert(requests, promise);
+            //future.get();
+
+            // We also set the original IL function body, in case the revert did not succeeded.
+            // We do that just as a safeguard, it has harmless effect.
+            //const auto [pMethodBytes, methodSize] = FaultTolerantTracker::Instance()->GetILBodyAndSize(moduleId, methodId);
+            //pFunctionControl->SetILFunctionBody(methodSize, pMethodBytes);
+
+            return m_methodRewriter->Rewrite(moduleHandler, methodHandler, pFunctionControl);
+        }
+        else
+        {
+            return ApplyKickoffInstrumentation(moduleHandler, methodHandler, pFunctionControl);   
+        }
+    }
+    else if (FaultTolerantTracker::Instance()->IsOriginalMethod(moduleId, methodId))
+    {
+        return ApplyOriginalInstrumentation(moduleHandler, methodHandler, pFunctionControl);
+    }
+    else if (FaultTolerantTracker::Instance()->IsInstrumentedMethod(moduleId, methodId))
+    {
+        const auto methodIdOfKickoff =
+            FaultTolerantTracker::Instance()->GetKickoffMethodFromInstrumentedMethod(moduleId, methodId);
+        const auto [pMethodBytes, methodSize] =
+            FaultTolerantTracker::Instance()->GetILBodyAndSize(moduleId, methodIdOfKickoff);
+
+        pFunctionControl->SetILFunctionBody(methodSize, pMethodBytes);
+
+        // substitute the methodHandler of the instrumented duplication with the original (the one of the kickoff)
+        if (!moduleHandler->TryGetMethod(methodIdOfKickoff, &methodHandler))
+        {
+            Logger::Warn("FaultTolerantRewriter::RewriteInternal(): Failed to substitute the methodHandler of the instrumented duplication with the original's.");
+            return S_FALSE;
+        }
+
+        InjectSuccessfulInstrumentationLambda injectSuccessfulInstrumentation =
+            [this](RejitHandlerModule* moduleHandler, RejitHandlerModuleMethod* methodHandler,
+                   ICorProfilerFunctionControl* pFunctionControl, LPCBYTE pbILMethod) -> HRESULT {
+            return this->InjectSuccessfulInstrumentation(moduleHandler, methodHandler, pFunctionControl, pbILMethod);
+        };
+
+
+        auto faultTolerantFunctionControl = std::make_unique<fault_tolerant::FaultTolerantCorProfilerFunctionControl>(pFunctionControl, moduleId, methodId, moduleHandler, methodHandler, injectSuccessfulInstrumentation);
```

Comment:
Does the reference to `faultTolerantFunctionControl` create a memory leak? I'm not really familiar with how the new pointers in C++ operate so I might not be understanding this correctly, but I don't see where we free the new object

Reply:
At the end of the block (when the 'smart pointer' will be out of scope), this `unique_ptr` will release the memory it holds. In general speaking, `unique_ptr` takes ownership of the memory and when it is passed from one place to another, _ownership_ is transferred (effectively 'stealing' the underlying memory).

The following call to the `Rewrite` is getting the raw underlying pointer (faultTolerantFunctionControl.get()), but since the rewriter is a synchronous call there is no issue with premature memory release before that call ends.

Reply:
> At the end of the block (when the 'smart pointer' will be out of scope), this unique_ptr will release the memory it holds.

Ah, this concept is new to me (can you tell I barely write C++ 😛). Thanks for explaining that to me!